File: 2015/day02.fs
\ gforth day02.fs -e part# <.input-day02
\ add an extra <newline> to input for `key` to detect

\ == part1 =====================================================
\ read 0+ digits and 1 non-digit ('x' or <newline>), put number on TOS
: getInteger ( -- n )
    0 begin key
       dup [char] 0 < if drop exit then
       dup [char] 9 > if drop exit then
           [char] 0 - swap 10 * + again ;

\ assume well-formed input LxWxH<newline>...<newline><newline>
\ and no presents with 0 side
: getLWH ( -- L W H | -- 0 )
   getInteger dup 0= if exit then \ blank final line
   getInteger getInteger ;

\ surface of box + area of smallest side
: paper ( L W H -- paper-needed )
   sort3asc 2dup * >r rot tuck * >r * 3 * r> r> + 2 * + ;

: part1
   0 begin getLWH dup 0 <> while paper + repeat drop
   cr ." Q: How many total square feet of wrapping paper should they order?"
   cr ." A: They should order " . ." square feet of wrapping paper." cr cr ;

\ == part2 =====================================================
\ smallest perimeter on any face + volume of box
: ribbon ( L W H -- ribbon-needed )
   sort3desc 2dup + 2 * >r * * r> + ;

: part2
   0 begin getLWH dup 0 <> while ribbon + repeat drop
   cr ." Q: How many total feet of ribbon should they order?"
   cr ." A: They should order " . ." feet of ribbon." cr cr ;

File: 2015/day02.c
#include <stdio.h>
#include <stdlib.h>

unsigned long long wneeded(char *box) {
    unsigned long l, w, h, top;
    // assume string is well-formatted, no overflows
    l = strtoul(box, &box, 10); box++; // skip 'x'
    w = strtoul(box, &box, 10); box++; // skip 'x'
    h = strtoul(box, &box, 10);
    top = l;
    if (w > top) top = w;
    if (h > top) top = h;
    unsigned long long total = l * w * h / top;
    total += 2 * l * w;
    total += 2 * l * h;
    total += 2 * w * h;
    return total;
}

unsigned long long part1(char (*data)[40], unsigned n) {
    unsigned long long total = 0;
    for (unsigned k = 0; k < n; k++) {
        total += wneeded(data[k]);
    }
    return total;
}

unsigned long long rneeded(char *box) {
    unsigned long l, w, h, top;
    // assume string is well-formatted, no overflows
    l = strtoul(box, &box, 10); box++; // skip 'x'
    w = strtoul(box, &box, 10); box++; // skip 'x'
    h = strtoul(box, &box, 10);
    top = l;
    if (w > top) top = w;
    if (h > top) top = h;
    unsigned long long total = l + w + h - top;
    total *= 2;
    total += l * w * h;
    return total;
}

unsigned long long part2(char (*data)[40], unsigned n) {
    unsigned long long total = 0;
    for (unsigned k = 0; k < n; k++) {
        total += rneeded(data[k]);
    }
    return total;
}

int main(void) {
    char data[2000][40];
    unsigned ndata = 0;
    while (fgets(data[ndata], 40, stdin)) {
        ndata++;
    }
    printf("DAY02-PART1: order %llu square feet of wrapping paper.\n", part1(data, ndata));
    printf("DAY02-PART2: order %llu feet of ribbon.\n", part2(data, ndata));
}
